rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,1039 @@
+  assert(count);
+
+  *count = std::count_if(
+    entities.begin(),
+    entities.end(),
+    [&topic_name](const GraphCache::EntityGidToInfo::value_type & elem)
+    {
+      return elem.second.topic_name == topic_name;
+    });
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+GraphCache::get_writer_count(
+  const std::string & topic_name,
+  size_t * count) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  if (!count) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  return __get_count(data_writers_, topic_name, count);
+}
+
+rmw_ret_t
+GraphCache::get_reader_count(
+  const std::string & topic_name,
+  size_t * count) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  if (!count) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  return __get_count(data_readers_, topic_name, count);
+}
+
+enum class EndpointCreator
+{
+  ROS_NODE = 0,
+  UNDISCOVERED_ROS_NODE = 1,
+  BARE_DDS_PARTICIPANT = 2,
+};
+
+static
+std::tuple<std::string, std::string, EndpointCreator>
+__find_name_and_namespace_from_entity_gid(
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  rmw_gid_t participant_gid,
+  rmw_gid_t entity_gid,
+  bool is_reader)
+{
+  auto it = participant_map.find(participant_gid);
+  if (participant_map.end() == it) {
+    return {"", "", EndpointCreator::BARE_DDS_PARTICIPANT};
+  }
+  for (const auto & node_info : it->second.node_entities_info_seq) {
+    auto & gid_seq = is_reader ? node_info.reader_gid_seq : node_info.writer_gid_seq;
+    auto it = std::find_if(
+      gid_seq.begin(),
+      gid_seq.end(),
+      [&](const rmw_dds_common::msg::Gid & gid) -> bool {
+        return 0u == std::memcmp(gid.data.data(), entity_gid.data, RMW_GID_STORAGE_SIZE);
+      }
+    );
+    if (gid_seq.end() != it) {
+      return {node_info.node_name, node_info.node_namespace, EndpointCreator::ROS_NODE};
+    }
+  }
+  return {"", "", EndpointCreator::UNDISCOVERED_ROS_NODE};
+}
+
+using DemangleFunctionT = GraphCache::DemangleFunctionT;
+
+static
+rmw_ret_t
+__get_entities_info_by_topic(
+  const GraphCache::EntityGidToInfo & entities,
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  const std::string & topic_name,
+  DemangleFunctionT demangle_type,
+  bool is_reader,
+  rcutils_allocator_t * allocator,
+  rmw_topic_endpoint_info_array_t * endpoints_info)
+{
+  assert(allocator);
+  assert(endpoints_info);
+
+  if (0u == entities.size()) {
+    return RMW_RET_OK;
+  }
+
+  size_t size = std::count_if(
+    entities.begin(),
+    entities.end(),
+    [&topic_name](const GraphCache::EntityGidToInfo::value_type & item) {
+      return item.second.topic_name == topic_name;
+    }
+  );
+  if (0u == size) {
+    return RMW_RET_OK;
+  }
+
+  rmw_ret_t ret = rmw_topic_endpoint_info_array_init_with_size(
+    endpoints_info,
+    size,
+    allocator);
+  // TODO(ivanpauno): `count` should be initialized by
+  //  `rmw_topic_endpoint_info_array_init_with_size`, and should be probably called `size`.
+  endpoints_info->count = size;
```

Comment:
@ivanpauno I think this needs an update.

Reply:
Done, see https://github.com/ros2/rmw_dds_common/pull/4/commits/6247fc72227f2f483682673f714e381c3213a936.